### ISSUE TITLE
tagged subsets of elements

### ISSUE THREAD

**Issue by gdmcbain** [nschloe/meshio#290]
The arrival of MSH4 #280 with Gmsh 4 broke some applications downstream of meshio (e.g. nschloe/pygmsh#209, kinnala/scikit-fem#41) but also included innovation in the treatment of ‘entities’ and ‘physical tags’ which prompts a review of the discussion of how to implement tagged subsets of elements, a.k.a. 'regions' #175, which goes beyond this single format or format version.

While #280 was closed once meshio was able to read the nodes and elements #283 #287, and data defined thereon #288, `meshio.read` does not read the physical tags from MSH4 as it did from the now legacy version of Gmsh's native format, MSH2.  These physical tags belong to elements (of various dimension from zero up to that of the domain) and are used in applications to define subdomains (materials, switching terms), mixed boundary conditions, zones for measurements, etc.  For an elementary example in which the Laplace equation is solved in a plane region with boundary divided into two parts with specified different uniform potentials and the rest insulation, see [gh:kinnala/scikit-fem/examples/ex13.py](https://github.com/kinnala/scikit-fem/blob/master/examples/ex13.py).

As more installations of Gmsh are upgraded to v. 4, repeats of breakages like those of @dajuno in #37 

> Can I expect the converter to handle gmsh's physical groups?

@gpkc in #43 

> This is actually very important to gmsh users. It's a very straightforward way to set boundary conditions, for example. This should not be neglected probably.
 
@cbcoutinho  in #169 

> I use gmsh physical groups pretty extensively, so it was a surprise to me when all of my tests started failing with an updated version of meshio.

(although here it would be an updated Gmsh rather than meshio) or @dalcinl in #197 

>  A few months back, regions were encoded in output Gmsh files just fine. Now they are not, 

(although this was about `meshio.write` rather than `meshio.read`).

There was a lengthy discussion on how to implement regions in #175 which was closed with the tags going into `cell_data[type]['gmsh:physical']` in #199, but this may need revisiting; in reviewing some sketchy exploration of the physical tags in MSH4 #289 , @nschloe wrote:
> we have to make a decision on how to use the tag groups in MSH4. I'm not sure about that yet. We used to simply create integer node/element data, but really that's quite ineffective. Perhaps we can do something that is common to MSH4 and Exodus which also has node/element sets. (And another mesh format, perhaps Medit?)

This issue is launched to discuss how to use the tags in MSH4 in a way which is common to other formats.

**Comment by gdmcbain** [comment]
With regard to MEDIT's `mesh` format, as defined in §7.2.1 ‘The `mesh` format’ of

* Pascal Frey. MEDIT : An interactive mesh visualization software. RT-0253, INRIA. 2001, pp. 41. [inria-00069921](https://hal.inria.fr/inria-00069921/fr/)

attributes one _ref_ to the syntax for each of the `Vertices`, `Edges`, `Triangles`, `Quadrilaterals`, `Tetrahedra`, and `Hexaedra`.  It doesn't say what this _ref_ is, but I assume it's an integer tag, which is how it's interpreted in _FreeFem++_, for example.

Gmsh writes this with `-format mesh`.  By default (and originally only), it used the elementary (geometric) entity for _ref_, but about a decade ago it got the option `Mesh.SaveElementTagType=2` to use the physical tag; see, e.g., [[Gmsh] Format Mesh and Boundary Conditions](http://onelab.info/pipermail/gmsh/2009/005057.html) (2009-12-03):

> > Maybe there exists a option that allows me to take those physical tags into account ?

> Yes, this was added recently (v. 2.4.2, I think): `Mesh.SaveElementTagType=2`

For example, [t1.mesh.gz](https://github.com/nschloe/meshio/files/2334368/t1.mesh.gz) was produced by
```shell
gmsh -2 -format mesh -string 'Mesh.SaveElementTagType=2;' t1.geo
```

In short: looks like exactly one `int` tag for each node and each element.  

That would map straightforwardly to `node_data` and `cell_data`, although a name for the key is required.  For the `cell_data`, the tag is exactly the same as what used to go in `gmsh:physical` (assuming `Mesh.SaveElementTagType=2`) but that's a funny name for a key that might have nothing to do with Gmsh.

**Comment by tianyikillua** [comment]
With the help of @gdmcbain I'm now working on reading/writing node/elements sets in a MED file.

MED format allows tagging a same node/element with several sets: a node/element can belong to several different node/element sets, while still being able to define

- a `point id -> unique node set id` map, à la `point_data`
- a `cell id -> unique element set id` map, à la `cell_data`

This is achieved by using **unions** of sets. If node 1 belongs to both `A` and `B`, and a new set `A-B` is internally created, see https://github.com/nschloe/meshio/issues/347#issuecomment-474251078.

In the first time, I will construct these two `data` to be contained in `point_data` and in `cell_data`. Maybe we could now find a neural name for these two data. Currently we have `gmsh:physical` for the latter, which is not neural at all.

Maybe we could follow the logic of FEniCS and name these two fields

- `PointFunction`
- `CellFunction` (even though this function is deprecated in FEniCS in favor of the generic `MeshFunction`)

and also in analogy to `points`, `cells`, `point_data` and `cell_data`?

**Comment by gdmcbain** [comment]
> This is achieved by using **unions** of sets. If node 1 belongs to both `A` and `B`, and a new set `A-B` is internally created

This sounds more like the set-theoretic _intersection_ than union.  The MED specification doesn't call it a ‘union’, does it?

In general, I think it's easy enough to convert between the two kinds of representation of tagged subsets of elements using this kind of reasoning, yes.

A single subset, given as a collection of indices of cells of a given type can be converted to an indicator array; conceptually: `array([i in subset for i in range(len(cell[type]))])`.

Given a number of such subsets, one could form the indicator array for each and then combine them as bits; conceptually, `2**arange(len(indicators)) @ stack(indicators)`.  I think that that gives a map _element_ &rarr; _region_ where the regions are defined by all the intersections of subsets, i.e. given which of these regions that a cell belongs to, one can deduce which of the subsets it belongs to by decomposing the integer label of the region into bits.

Conversely, a subset as a collection of indices can be obtained from a mapping _element_ &rarr; _region_ using something like `nonzero(regions == region_id)[0]`.

These aren't necessarily how these conversions should be implemented, just sketches for a proof of concept.

**Comment by tianyikillua** [comment]
Yes intersections!!!

**Comment by gdmcbain** [comment]
> Maybe we could follow the logic of FEniCS and name these two fields

   * `PointFunction`, or `point_function`
    * `CellFunction`, or `cell_function` (even though this function is deprecated in FEniCS in favor of the generic `MeshFunction`)

Yes, that's pretty good.  Maybe is there a word instead of function that implies that the function is to something discrete rather than continuous?  Sometimes people talk about colour, as in [graph colouring](https://en.wikipedia.org/wiki/Graph_coloring), but I read there that that's a special case of [graph labelling](https://en.wikipedia.org/wiki/Graph_labeling).  So how about ‘label’; e.g. `point_label` and `cell_label`?  (The MEDIT documentation uses ‘ref’.)

What about going the other way?  ‘Collections’, ‘regions’, ‘subsets’, ‘sets’?  Any of those would be O. K., I think.  ‘Group’ has an unrelated meaning in algebra.  Set theory seems relevant.

**Comment by tianyikillua** [comment]
So we'd like to give a name to the following map, that associates to every point/cell a unique set id.

1. `point/cell_function`
2. `tags` (we can use the same name in fact for both of them!)
3. `sets`
4. ...

Maybe we should do a vote. I prefer now `tags`, simple and explains the discrete nature of the map. If the unique id for node sets and for element sets are different, I can also merge my `point_tags` and `cell_tags` (or your `field_data`), so that we have

``` python
mesh.tags = {unique node/element set id: [set names]}
```

**Comment by gdmcbain** [comment]
> I prefer now tags, simple and explains the discrete nature of the map.

+1

For ‘tag’.

The ‘map’ is a map in the mathematical sense but not in the Pythonic sense.  (Ditto for ‘function' which is a mathematical synonym but means something different again in Python.)  Dunno whether that matters.  I don't have a  better proposal at the moment.

**Comment by gdmcbain** [comment]
```python
mesh.point_data["tags"] = {node id: unique node set id}
mesh.cell_data[cell_type]["tags] = {element id: unique element set id}
```

I'm not sure that I understand the right-hand sides here.  Does 'node id' run over `numpy.arange(len(mesh.points))` and `element id` over `numpy.arange(len(mesh.cells[cell_type]))`?

**Comment by tianyikillua** [comment]
Oh yes

```
mesh.point_data["tags"] = np.array([corresponding unique node set id's])
mesh.cell_data[cell_type]["tags] = np.array([unique element set id's])
```

**Comment by gdmcbain** [comment]
Was there any significance in the tags being _negative_ integers in Remi's example in #347?

**Comment by gdmcbain** [comment]
> `mesh.point_data["tags"] = np.array([corresponding unique node set id's], type=int)`

So that's fine (and ditto for `cell_data[cell_type]["tags"]`)

An alternative would be, e.g.:

```
mesh.point_data["tags"] = {"a tag": an_array, "another tag": another_array, …}
```
where each value on the right-hand side has type int and shape `(len(mesh.points),)`, and one could have as many tags as one wished.

Not sure which is better, just floating the idea.  I think that they are ultimately equivalent (isomorphic), just not sure whether there are advantages of convenience, efficiency, etc.

**Comment by tianyikillua** [comment]
> Was there any significance in the tags being _negative_ integers in Remi's example in #347?

See the tags in https://github.com/nschloe/meshio/pull/352, it seems that all node sets are tagged with positive id's, while element sets with negative id's.

**Comment by tianyikillua** [comment]
> An alternative would be, e.g.:
> 
> ```
> mesh.point_data["tags"] = {"a tag": an_array, "another tag": another_array, …}
> ```
> where each value on the right-hand side has type int and shape `(len(mesh.points),)`, and one could have as many tags as one wished.
> 
> Not sure which is better, just floating the idea. I think that they are ultimately equivalent (isomorphic), just not sure whether there are advantages of convenience, efficiency, etc.

With this we could store many ways to tag node/elements, however in a mesh there should be only one set of tags...?

**Comment by gdmcbain** [comment]
> With this we could store many ways to tag node/elements, however in a mesh there should be only one set of tags...?

Not necessarily.  I think you gave an [example](https://github.com/nschloe/meshio/issues/347#issuecomment-474252859) of multiple tags in #347.  

Christophe Geuzaine gave another on [[Gmsh] ‘node and element sparsity in msh4 ’](http://onelab.info/pipermail/gmsh/2018/012432.html) (2018-09-15):
> All applications where you have overlapping groups: functional characterization ("left wing") vs. materials ("aluminum" and "carbon fiber"), local feature ("wire 2") vs. global one ("stator winding"), overlapping boundary conditions for multiphysics applications ("zero flux" vs "fixed displacement"), overlapping boundaries of adjacent parts (sharing some entities), etc.

As I say though, I think these are ultimately isomorphic; it's just questions of convenience, efficiency, clarity, &c.

**Comment by gdmcbain** [comment]
Another thought: at the moment, `point_data` and `cell_data` contain both discrete tags (like "medit:ref" and "gmsh:physical") and float fields (typically finite-element solutions).  I think it was the latter that they were originally created for.  What about identifying anything in `cell_data[cell_type]` with a discrete `dtype` as a tag?  Probably a silly idea.  I'm not sure that NumPy has a predicate is_discrete for dtypes and I wouldn't know how to test a dtype for discreteness.  Probably your idea of a key "tags" is simpler and clearer (whether the value is a single array or itself a dict).

**Comment by gdmcbain** [comment]
Suggested principle : defer conversion between equivalent internal representations whenever and as long as possible.

**Comment by tianyikillua** [comment]
> > With this we could store many ways to tag node/elements, however in a mesh there should be only one set of tags...?
> 
> Not necessarily. I think you gave an [example](https://github.com/nschloe/meshio/issues/347#issuecomment-474252859) of multiple tags in #347.
> 
> Christophe Geuzaine gave another on [[Gmsh] ‘node and element sparsity in msh4 ’](http://onelab.info/pipermail/gmsh/2018/012432.html) (2018-09-15):
> 
> > All applications where you have overlapping groups: functional characterization ("left wing") vs. materials ("aluminum" and "carbon fiber"), local feature ("wire 2") vs. global one ("stator winding"), overlapping boundary conditions for multiphysics applications ("zero flux" vs "fixed displacement"), overlapping boundaries of adjacent parts (sharing some entities), etc.
> 
> As I say though, I think these are ultimately isomorphic; it's just questions of convenience, efficiency, clarity, &c.

Oh, I see. You actually mean tags with a non empty intersection. We could just group them together in one tag set, just increasing the number of tags in this set.

**Comment by ruizbaier** [comment]
Hi, related also to #175, I have several Exodus meshes with surface labels and I need to export them to some format readable by GMSH. I can still do this with meshio (to e.g. .vtk or to .msh) but all labels are lost. Is there a way of doing this still with meshio? I've tried also dolfin-convert but the ordering of the nodes in the resulting ncdf files (after conversion from .e files) is wrong. 
Is this the correct place to ask this question? Thank you so much!

**Comment by gdmcbain** [comment]
A similar difficulty was reported in importing MED in #389; there it was written that only by using Gmsh (instead of meshio) could the tagged subsets of elements be retained.  This does seem unfortunate, and a different outcome was reported in #347 but the handling of tagged subsets of elements by meshio does seem to be quite different for MSH, MEDIT, and MED formats; I don't know about Exodus.

**Comment by gdmcbain** [comment]
Fixed by #699, at least for Gmsh's MSH 4.1.
